resources/type-system/inference.md

```diff
@@ -0,0 +1,830 @@
+  - if `S` is `U Function<X0 extends B0, ...., Xk extends Bk>(T0 x0, ...., Tn xn, {Tn+1 xn+1, ..., Tm xm})`
+      the union of:
+    - the covariant occurrences of `T` in `U`
+    - the contravariant occurrences of `T` in `Ti` for `i` in `0, ..., m`
+
+The contravariant occurrences of a type `T` in another type `S` are:
+  - if `S` is `Future<U>`
+    - the contravariant occurrences of `T` in `U`
+  - if `S` is `FutureOr<U>`
+    - the contravariant occurrencs of `T` in `U`
+  - if `S` is an interface type `C<T0, ..., Tk>`
+    - the union of the contravariant occurrences of `T` in `Ti` for `i` in `0, ..., k`
+  - if `S` is `U Function<X0 extends B0, ...., Xk extends Bk>(T0 x0, ...., Tn xn, [Tn+1 xn+1, ..., Tm xm])`, 
+      the union of:
+    - the contravariant occurrences of `T` in `U`
+    - the covariant occurrences of `T` in `Ti` for `i` in `0, ..., m`
+  - if `S` is `U Function<X0 extends B0, ...., Xk extends Bk>(T0 x0, ...., Tn xn, {Tn+1 xn+1, ..., Tm xm})`
+      the union of:
+    - the contravariant occurrences of `T` in `U`
+    - the covariant occurrences of `T` in `Ti` for `i` in `0, ..., m`
+
+The invariant occurrences of a type `T` in another type `S` are:
+  - if `S` is `Future<U>`
+    - the invariant occurrences of `T` in `U`
+  - if `S` is `FutureOr<U>`
+    - the invariant occurrencs of `T` in `U`
+  - if `S` is an interface type `C<T0, ..., Tk>`
+    - the union of the invariant occurrences of `T` in `Ti` for `i` in `0, ..., k`
+  - if `S` is `U Function<X0 extends B0, ...., Xk extends Bk>(T0 x0, ...., Tn xn, [Tn+1 xn+1, ..., Tm xm])`, 
+      the union of:
+    - the invariant occurrences of `T` in `U`
+    - the invariant occurrences of `T` in `Ti` for `i` in `0, ..., m`
+    - all occurrences of `T` in `Bi` for `i` in `0, ..., k`
+  - if `S` is `U Function<X0 extends B0, ...., Xk extends Bk>(T0 x0, ...., Tn xn, {Tn+1 xn+1, ..., Tm xm})`
+      the union of:
+    - the invariant occurrences of `T` in `U`
+    - the invariant occurrences of `T` in `Ti` for `i` in `0, ..., m`
+    - all occurrences of `T` in `Bi` for `i` in `0, ..., k`
+
+### Type schema elimination (least and greatest closure of a type schema)
+
+We define the least closure of a type schema `P` with respect to `?` to be `P`
+with every covariant occurrence of `?` replaced with `Null`, and every invariant
+or contravariant occurrence of `?` replaced with `Object`.
+
+We define the greatest closure of a type schema `P` with respect to `?` to be
+`P` with every invariant and covariant occurrence of `?` replaced with `Object`,
+and every contravariant occurrence of `?` replaced with `Null`.
+
+Note that the closure of a type schema is a proper type.
+
+Note that the least closure of a type schema is always a subtype of any type
+which matches the schema, and the greatest closure of a type schema is always a
+supertype of any type which matches the schema.  **This is not true for invariant
+types.**
+
+TODO: decide what to do about invariant types.
```

Comment:
Indeed. I'll return this review now rather than going into a couple of more tricky areas (like this one), and then we can discuss it later on.